File: tui/src/protocol.ts
/**
 * Protocol type definitions for the acomm bridge.
 * Mirrors Rust's ProtocolEvent with serde external tagging.
 *
 * Example JSON from the bridge:
 *   {"Prompt":{"text":"hello","provider":null,"channel":"tui"}}
 *   {"AgentChunk":{"chunk":"...","channel":null}}
 *   {"AgentDone":{"channel":null}}
 *   {"ProviderSwitched":{"provider":"Gemini"}}
 */

export type AgentProvider = 'Gemini' | 'Claude' | 'Codex' | 'OpenCode' | 'Mock';

export const AGENT_PROVIDERS: AgentProvider[] = ['Gemini', 'Claude', 'Codex', 'OpenCode'];

/** Available models for each provider. */
export const PROVIDER_MODELS: Record<AgentProvider, string[]> = {
  Gemini:   ['gemini-2.5-flash', 'gemini-2.5-pro', 'gemini-2.0-flash'],
  Claude:   ['claude-opus-4-6', 'claude-sonnet-4-6', 'claude-haiku-4-5'],
  Codex:    ['gpt-4o', 'gpt-4o-mini', 'o1-mini'],
  OpenCode: ['default'],
  Mock:     ['mock-model'],
};

export type ProtocolEvent =
  | { Prompt: { text: string; provider: AgentProvider | null; channel: string | null } }
  | { AgentChunk: { chunk: string; channel: string | null } }
  | { AgentDone: { channel: string | null } }
  | { SystemMessage: { msg: string; channel: string | null } }
  | { StatusUpdate: { is_processing: boolean; channel: string | null } }
  | { BridgeSyncDone: {} }
  | { SyncContext: { context: string } }
  | { ProviderSwitched: { provider: AgentProvider } }
  | { ModelSwitched: { model: string } };

/** Returns the variant name of a ProtocolEvent. */
export function eventKind(event: ProtocolEvent): string {
  return Object.keys(event)[0]!;
}

/** Converts an AgentProvider to its CLI command name (lowercase). */
export function providerCommandName(provider: AgentProvider): string {
  return provider.toLowerCase();
}

/** Case-insensitively finds the valid AgentProvider for a given string. */
export function normalizeProvider(name: string | null | undefined): AgentProvider {
  if (!name) return 'Gemini';
  const target = name.toLowerCase();
  for (const provider of AGENT_PROVIDERS) {
    if (provider.toLowerCase() === target) return provider;
  }
  if ('mock'.toLowerCase() === target) return 'Mock';
  return 'Gemini';
}

/** Safely gets the model list for a provider, falling back to Gemini if the provider is invalid. */
export function getModelsForProvider(provider: AgentProvider | string | null | undefined): string[] {
  const normalized = normalizeProvider(provider as any);
  return PROVIDER_MODELS[normalized] ?? PROVIDER_MODELS['Gemini'];
}

File: src/protocol.rs
use acore::AgentProvider;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ProtocolEvent {
    Prompt { 
        text: String, 
        provider: Option<AgentProvider>,
        channel: Option<String>,
    },
    /// エージェントからの回答の断片（チャンク）。
    AgentChunk { 
        chunk: String,
        channel: Option<String>,
    },
    AgentDone {
        channel: Option<String>,
    },
    SystemMessage { 
        msg: String,
        channel: Option<String>,
    },
    StatusUpdate { 
        is_processing: bool,
        channel: Option<String>,
    },
    SyncContext { context: String },
    ProviderSwitched { provider: AgentProvider },
    ModelSwitched { model: String },
}

impl ProtocolEvent {
    pub fn clone_channel(&self) -> Option<String> {
        match self {
            ProtocolEvent::Prompt { channel, .. } => channel.clone(),
            ProtocolEvent::AgentChunk { channel, .. } => channel.clone(),
            ProtocolEvent::AgentDone { channel, .. } => channel.clone(),
            ProtocolEvent::SystemMessage { channel, .. } => channel.clone(),
            ProtocolEvent::StatusUpdate { channel, .. } => channel.clone(),
            ProtocolEvent::SyncContext { .. }
            | ProtocolEvent::ProviderSwitched { .. }
            | ProtocolEvent::ModelSwitched { .. } => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::ProtocolEvent;
    use acore::AgentProvider;

    #[test]
    fn prompt_deserializes_provider_field() {
        let json = r#"{"Prompt":{"text":"hello","provider":"Gemini","channel":"tui"}}"#;
        let event: ProtocolEvent = serde_json::from_str(json).unwrap();
        match event {
            ProtocolEvent::Prompt { provider, .. } => {
                assert_eq!(provider, Some(AgentProvider::Gemini));
            }
            _ => panic!("expected Prompt"),
        }
    }

    #[test]
    fn provider_switched_serializes_provider_field() {
        let event = ProtocolEvent::ProviderSwitched { provider: AgentProvider::Claude };
        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains(r#""provider":"Claude""#));
        assert!(!json.contains(r#""tool":"Claude""#));
    }

    #[test]
    fn provider_switched_deserializes_provider_field() {
        let json = r#"{"ProviderSwitched":{"provider":"Codex"}}"#;
        let event: ProtocolEvent = serde_json::from_str(json).unwrap();
        match event {
            ProtocolEvent::ProviderSwitched { provider } => {
                assert_eq!(provider, AgentProvider::Codex);
            }
            _ => panic!("expected ProviderSwitched"),
        }
    }
}
